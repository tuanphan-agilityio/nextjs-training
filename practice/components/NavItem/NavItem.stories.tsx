import { Meta, Story } from '@storybook/react';

import NavItem, { NavItemProps } from '.';

export default {
  component: NavItem,
  title: 'NavItem',
  decorators: [
    (Story) => (
      <ul className='flex font-secondary-regular text-md'>
        <Story />
      </ul>
    ),
  ],
} as Meta;

const Template: Story<NavItemProps> = (args) => <NavItem {...args} />;

export const ActiveItem = Template.bind({});
ActiveItem.args = {
  isActive: true,
  href: '/active-link',
  title: 'Active Link',
};

export const InactiveItem = Template.bind({});
InactiveItem.args = {
  isActive: false,
  href: '/inactive-link',
  title: 'Inactive Link',
};
