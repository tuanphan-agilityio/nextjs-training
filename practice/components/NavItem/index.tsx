import Link from 'next/link';
import { FC, memo } from 'react';

interface NavItemProps {
  isActive: boolean;
  href: string;
  title: string;
}

const NavItem: FC<NavItemProps> = ({ href, title, isActive }) => {
  return (
    <li className='relative'>
      <Link href={href}>
        <a
          className={
            isActive
              ? 'after:content-[""] after:absolute after:bottom-0 after:left-1/2 after:w-1 after:h-1 after:rounded-full after:bg-quaternary after:transform after:-translate-x-1/2 text-quaternary font-secondary-bold'
              : 'hover:text-quaternary'
          }
        >
          {title}
        </a>
      </Link>
    </li>
  );
};

export type { NavItemProps };

export default memo(NavItem);
