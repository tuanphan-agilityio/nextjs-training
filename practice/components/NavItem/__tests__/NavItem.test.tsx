import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import NavItem from '..';

describe('NavItem Component', () => {
  const mockProps = {
    href: '/mock-link',
    title: 'Mock Link',
  };

  it('renders correctly with mockProps and is not active', () => {
    const { container } = render(<NavItem {...mockProps} isActive={false} />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly when active', () => {
    const { container } = render(<NavItem isActive={true} {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('displays the correct title', () => {
    render(<NavItem {...mockProps} isActive={false} />);
    const link = screen.getByText('Mock Link');
    expect(link).toBeInTheDocument();
  });

  it('contains the correct href link', () => {
    render(<NavItem {...mockProps} isActive={false} />);
    const link = screen.getByRole('link', { name: /Mock Link/i });
    expect(link).toHaveAttribute('href', '/mock-link');
  });

  it('applies active class when active', () => {
    render(<NavItem isActive={true} {...mockProps} />);
    const link = screen.getByText('Mock Link');
    expect(link).toHaveClass('after:bg-quaternary');
  });

  it('applies hover class when not active', () => {
    render(<NavItem isActive={false} {...mockProps} />);
    const link = screen.getByText('Mock Link');
    userEvent.hover(link);
    expect(link).toHaveClass('hover:text-quaternary');
  });
});
