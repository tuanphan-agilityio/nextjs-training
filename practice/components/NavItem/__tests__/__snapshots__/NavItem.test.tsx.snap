// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem Component renders correctly when active 1`] = `
<div>
  <li
    class="relative"
  >
    <a
      class="after:content-[""] after:absolute after:bottom-0 after:left-1/2 after:w-1 after:h-1 after:rounded-full after:bg-quaternary after:transform after:-translate-x-1/2 text-quaternary font-secondary-bold"
      href="/mock-link"
    >
      Mock Link
    </a>
  </li>
</div>
`;

exports[`NavItem Component renders correctly with mockProps and is not active 1`] = `
<div>
  <li
    class="relative"
  >
    <a
      class="hover:text-quaternary"
      href="/mock-link"
    >
      Mock Link
    </a>
  </li>
</div>
`;
